rfcs/000-consolidate-amundsen-microservices.md

```diff
@@ -0,0 +1,154 @@
+Python packages are pretty much the same for each proxy, which makes it hard to sync across all the above repositories. As a result most of the time, each proxy is running its own version of the dependency. 
+
+A few examples:
+
+| Package             | amundsenmetadata   | amundsenfrontend  |
+| ------------------  |:------------------:| -----------------:|
+| amundsen-common     | >=0.8.1            | ==0.6.0           |
+| flake8              | ==3.5.0            | ==3.8.4           |
+| flake8-tidy-imports | ==1.1.0            | ==4.2.1           |
+
+and many more... 
+
+*This change will put all the packages in just one place, making it easier to manage dependencies and no multiple variants of the same package.*
+
+#### 2: Development Efforts
+Having multiple repositories makes it really hard to implement a feature. Implementation and testing require efforts to synchronize and then code reviews, and finally, all the PRs across multiple repositories need to land in master at a certain time or at the same time. 
+
+*This change results in faster local development, one PR to fix a bug or a feature, no dependency hell as we are facing today, hence attract more contributors.*
+
+#### 3: Customization & Deployments
+One of the frequently mentioned topics re Amundsen is the complexity of the architecture, and the efforts required to customize Amundsen. Having multiple repositories results in multiple components to install, customize, manage and fix.
+
+*This change will make the customization much easier and the deployment will be pretty straightforward, with only 2 components. No multilevel docker files, easy maintenance of the infrastructure.* 
+
+Custom endpoints will still be injectable the same way we are doing today. We will write a blog/tutorial around each customization.
+
+#### 4: Code Duplication
+From requirements to config files, models, helper functions, exceptions handling, CI/CD pipelines, Docker files, and other repository management like licenses, PR/Issues templates, etc., is pretty much the same in all these microservices. Change in one repository deviates that repository from others at the moment. 
+
+*This change will move all the above into a single repository making it easy to change anything without keeping track of multiple repositories, and redundant code.* 
+
+Having one config file does NOT mean we can not have the incubator integration. It will be the same process as of today i.e., a separate incubation repo, which implements the required methods, and then install that as a requirement and set that as a proxy within the config file.
+
+
+## Transition Path
+Most of the code is already there and is spread across multiple repositories. I can see this happen in multiple phases.
+
+#### Phase I:
+Merge `amundsenmetadatalibrary` and `amundsensearchlibrary`. This should not take too long and will be easy to maintain the same URL patterns for each under separate directories within one repository. 
```

Comment:
Sure thing, 
There will not be a separate release of metadata and search after this phase. We will maintain only one config file and a Docker file. 

Since, we are not deleting the old services i.e., amundsenmetadatalibray, and amundsensearchlibrary, users can still migrate to the new architecture in parallel with the old system. Because the data is not changing at all, this is possible. 
That would allow you to set up the new monolith, and leave the other system running in parallel. then you would change the front-end to use the monolith as the metadata service and shut down the metadata microservice. then same for search service. This would give a path for people to test their deployment in a way that is de-risked. (I will also add this in the RFC doc :) )

Most of the configurations variable in our services are the same i.e., `statsd, logging, swagger, proxy credentials`. So combining those configs will not break anything. A single entry point and requirements will make it easier to have one docker image and PyPI release. 

config file will look something like this.

```python

METADATA_PROXY_CLIENTS = {
    'NEO4J': 'metadata.proxies.neo4j.proxy.Neo4jProxy',
    'ATLAS': 'metadata.proxies.atlas.proxy.AtlasProxy',
    'NEPTUNE': 'metadata.proxies.neptune.proxy.NeptuneGremlinProxy'
}

class MetadataConfig:
		METADATA_PROXY_HOST = os.environ.get('METADATA__PROXY_HOST', 'localhost')
    METADATA_PROXY_PORT = os.environ.get('METADATA__PROXY_PORT', 7687)
    METADATA_PROXY = os.environ.get('METADATA__PROXY', 'NEO4J')
		# For the users who would like to have custom proxy classes. This also enable us to install and use the custom proxies outside this repository
    METADATA_PROXY_CLIENT = os.environ.get('METADATA__PROXY_CLIENT', PROXY_CLIENTS[METADATA_PROXY])
    METADATA_PROXY_USER = os.environ.get('METADATA__PROXY_USER', 'neo4j')
		METADATA_PROXY_PASSWORD = os.environ.get('METADATA__PROXY_PASSWORD', 'test')

class SearchConfig:
    # SAME AS METADATA ABOVE

class Config(MetadataConfig, SearchConfig):
    LOG_FORMAT = '%(asctime)s.%(msecs)03d [%(levelname)s] %(module)s.%(funcName)s:%(lineno)d (%(process)d:' \
                 '%(threadName)s) - %(message)s'
    LOG_DATE_FORMAT = '%Y-%m-%dT%H:%M:%S%z'
    LOG_LEVEL = 'INFO'

    PROXY_ENCRYPTED = True
    """Whether the connection to the proxy should use SSL/TLS encryption."""

    # Prior to enable PROXY_VALIDATE_SSL, you need to configure SSL.
    # https://neo4j.com/docs/operations-manual/current/security/ssl-framework/
    PROXY_VALIDATE_SSL = False
    """Whether the SSL/TLS certificate presented by the user should be validated against the system's trusted CAs."""

    IS_STATSD_ON = False

    # Configurable dictionary to influence format of column statistics displayed in UI
    STATISTICS_FORMAT_SPEC: Dict[str, Dict] = {}

    SWAGGER_ENABLED = os.environ.get('SWAGGER_ENABLED', False)

    SWAGGER_TEMPLATE_PATH = os.path.join('api', 'swagger_doc', 'template.yml')
    SWAGGER = {
        'openapi': '3.0.2',
        'title': 'Metadata Service',
        'uiversion': 3
    }
```

Reply:
how could others continue contribute to source code without forcing to migrate first?

Reply:
I think fundamentally we are thinking it quite different. This RFC just makes Amundsen into a standalone Flask App with external plugin to swap storage while I think in long term the metadata service should evolve into a complex platform which could support different primitive like event triggering which the metadata could support different Apps .